Task2

patrikprovaznik opened the pull request.

patrikprovaznik committed:
In this assignment we created a client application that is making a requests to our server application from previous assignment.
For this assignment we used technologies like:
1. Dataclasses - We created dataclasses based od specific data of json received: name, description, type, ...
2. Using logger, requests and OOP - We created class that contains methods for example: class for getting data from server, class for analyzing what kind of data we received from server and also creating logs for these data and class for writing specific data into output file in json format.

patrikprovaznik committed:
In this assignment we created a client application that is making a requests to our server application from previous assignment.
For this assignment we used technologies like:
1. Dataclasses - We created dataclasses based od specific data of json received: name, description, type, ...
2. Using logger, requests and OOP - We created class that contains methods for example: class for getting data from server, class for analyzing what kind of data we received from server and also creating logs for these data and class for writing specific data into output file in json format.

patrikprovaznik committed:
In this assignment we created a client application that is making a requests to our server application from previous assignment.
For this assignment we used technologies like:
1. Dataclasses - We created dataclasses based od specific data of json received: name, description, type, ...
2. Using logger, requests and OOP - We created class that contains methods for example: class for getting data from server, class for analyzing what kind of data we received from server and also creating logs for these data and class for writing specific data into output file in json format.

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+from dataclasses_json import DataClassJsonMixin, config
+from marshmallow import fields, EXCLUDE
+
+# Logger
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
+
+file_handler = logging.FileHandler('output_folder/interfaces.log', mode='w')
+file_handler.setLevel(logging.INFO)
+file_handler.setFormatter(formatter)
+
+logger.addHandler(file_handler)
```
@patrikprovaznik put this to some util module that will take as an input (log_name: str, folder_path_name: str) that will be predefined and return the logger which you will use here

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+logger.addHandler(file_handler)
+
+
+@dataclass
+class AddressIpv4Inside(DataClassJsonMixin):
+    ip: str
+    netmask: str
+
+
+@dataclass
+class AddressIpv6Inside(DataClassJsonMixin):
+    ip: str
+    prefix_length: int = field(
+        metadata=config(mm_field=mm.fields.Number(data_key='prefix-length')))
+
+
+@dataclass
+class AddressIpv4(DataClassJsonMixin):
+    address: List[AddressIpv4Inside] = field(
+        metadata=config(mm_field=mm.fields.List(fields.Nested(AddressIpv4Inside.schema(unknown=mm.EXCLUDE)))),
+        default_factory=list)
+
+
+@dataclass
+class AddressIpv6(DataClassJsonMixin):
+    address: List[AddressIpv6Inside] = field(
+        metadata=config(mm_field=mm.fields.List(fields.Nested(AddressIpv6Inside.schema(unknown=mm.EXCLUDE)))),
+        default_factory=list)
+
+
+@dataclass
+class Interface(DataClassJsonMixin):
+    name: str
+    description: str = field(metadata=config(mm_field=mm.fields.String(data_key='description', load_default=str)))
+    type: str
+    enabled: bool
+    ip_address_1: AddressIpv4 = field(
+        metadata=config(mm_field=mm.fields.Nested(AddressIpv4.schema(unknown=mm.EXCLUDE), data_key='ietf-ip:ipv4')))
+    ip_address_2: AddressIpv6 = field(
+        metadata=config(mm_field=mm.fields.Nested(AddressIpv6.schema(unknown=mm.EXCLUDE), data_key='ietf-ip:ipv6')))
```
@patrikprovaznik put dataclasses to separate file data_type.py

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+        metadata=config(mm_field=mm.fields.Number(data_key='prefix-length')))
+
+
+@dataclass
+class AddressIpv4(DataClassJsonMixin):
+    address: List[AddressIpv4Inside] = field(
+        metadata=config(mm_field=mm.fields.List(fields.Nested(AddressIpv4Inside.schema(unknown=mm.EXCLUDE)))),
+        default_factory=list)
+
+
+@dataclass
+class AddressIpv6(DataClassJsonMixin):
+    address: List[AddressIpv6Inside] = field(
+        metadata=config(mm_field=mm.fields.List(fields.Nested(AddressIpv6Inside.schema(unknown=mm.EXCLUDE)))),
+        default_factory=list)
```
@patrikprovaznik these are extra unneeded classes.. you can do list of AddressIpv6Inside directly in Interface dataclass..

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+
+
+class InterfaceManager:
+    def __init__(self, url="http://127.0.0.1:5000/get-all-interfaces"):
```
@patrikprovaznik you shouldn t hold here the whole url.. that is not generic enough for future use. idealy it should contain protocol (https or https), host (ip or hostname) and port (integer) out of which you will construct url in init and keep only that there and all three can be in future configurable from config file

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+class InterfaceManager:
+    def __init__(self, url="http://127.0.0.1:5000/get-all-interfaces"):
+        self.api_url = url
+        self.interface_schema = Interface.schema()
```
@patrikprovaznik we don t need this here you can use it directly in the method where you need it

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+        with open('output_folder/logg_data.json', 'w') as f:
+            json.dump(self.analyze(), f, ensure_ascii=False, indent=4)
+
+
+interface_manager = InterfaceManager()
+interface_manager.write_all_data()
```
@patrikprovaznik create main which will call main method

something like

```
def main():
    interface_manager = InterfaceManager()
    interface_manager.write_all_data()


if __name__ == "__main__":
    main()
```
do you know why we do such?

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+    def __init__(self, url="http://127.0.0.1:5000/get-all-interfaces"):
+        self.api_url = url
+        self.interface_schema = Interface.schema()
+        self.parsed_data = self.get_all_interfaces()
```
@patrikprovaznik call this from analyze and from write all data inside. meaning we will have always the latest information

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+        self.interface_schema = Interface.schema()
+        self.parsed_data = self.get_all_interfaces()
+
+    def get_all_interfaces(self) -> List[Interface]:
```
@patrikprovaznik if you change according to above comment then please make this private

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+        parsed_interfaces = self.interface_schema.loads(json_data=response.text, many=True, unknown=EXCLUDE)
+        return parsed_interfaces
+
+    def analyze(self):
```
@patrikprovaznik return type

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+                'missing_completely_both': {'count': len(no_ip_addresses), 'interfaces': [no_ip_addresses]}}}
+        return logg_dict
+
+    def write_all_data(self):
```
@patrikprovaznik add return type

mirokovac commented on task2/InterfaceManager.py:
```diff
@@ -0,0 +1,109 @@
+from dataclasses import dataclass, field
+from typing import List
+
+import marshmallow as mm
+import requests
+from dataclasses_json import DataClassJsonMixin, config
+from marshmallow import fields, EXCLUDE
```
@patrikprovaznik do you have these in requirements file?

mirokovac replied:
@patrikprovaznik you are actually missing readme and requirements file.. please add that in

mirokovac reviewed (changes requested):
@patrikprovaznik .idea folder logg_data.json output_data.json put to gitignore.. each user should generate that for themselvs

patrikprovaznik committed:
adding utils.py argument_parser.py and data_type.py

patrikprovaznik committed:
1. adding config file flask_app_config.ini

patrikprovaznik committed:
adding argument parser so data can be loaded from arguments set by user for client side, creating separated file for dataclasses, also logger which one is configured from config file

patrikprovaznik committed:
removing output folder from this assignment, adding it to gitignore

mirokovac commented on task2/argument_parser.py:
```diff
@@ -0,0 +1,17 @@
+    parser.add_argument('-pt', '--path', dest='log_path', metavar='LOG_PATH',
+                        help="Set path with name for .log file, e.g. /home/Documents/<log_name>.log.", required=False,
+                        default="logs/logger_2.log")
+    parser.add_argument('-ll', '--level', dest='log_level', metavar='LOG_LEVEL', help="Set logging level",
```
@patrikprovaznik limit arguments to specific choices. use only INFO, DEBUG, WARNING, ERROR

mirokovac commented on task2/argument_parser.py:
```diff
@@ -0,0 +1,17 @@
+    parser.add_argument('--ssl', action='store_true', help='Set protocol for URL.')
+    parser.add_argument('-ho', '--host', dest='host', metavar='HOST', help='Set host for URL.', required=True)
+    parser.add_argument('-po', '--port', dest='port', metavar='PORT', help='Set port for URL.', required=True)
+    parser.add_argument('-nm', '--name', dest='log_name', metavar='LOG_NAME',
```
@patrikprovaznik this is not needed at all.. log_name should be based of the module that is being logged

mirokovac commented on task2/flask_app_config.ini:
```diff
@@ -0,0 +1,10 @@
+
+[general]
+log_path_1 = /home/patrik/Documents/patprotasks/task2/logs/logger_1.log
+log_name = logger_1
```
same as above

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,89 @@
+from utils import get_logger
+
+# Argument parser
+args = arg_pars()
```
@patrikprovaznik why is this outside main?

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,89 @@
+        raise SystemExit(e)
+
+
+# Setting URL
+set_url = f'{protocol(args.ssl)}://{args.host}:{args.port}/get-all-interfaces'
+url_check(set_url)
```
@patrikprovaznik  why is this outside of main?

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,89 @@
+                missing_ipv6.append(data.name)
+            if data.enabled:
+                enabled_count += 1
+            types[data.type] = types.get(data.type, 0) + 1
```
@patrikprovaznik I like this but try and use defaultdict from collections library..

mirokovac reviewed (commented):
@patrikprovaznik please clean up your .gitignore file.. anything from .idea should not be there

patrikprovaznik committed:
we did some adjustments: 1. adding logger, argument parser and URL setting into main function
2. we used defaultdict from collections library to count different types interfaces

patrikprovaznik committed:
1. deleting config file flask_app_config.ini, because we are using it in next task (task3)
2. adding argument parser, url and logger into main function
3. using __name__ for logger, to display the name of module that is being logged
4. auto creating folders for output data and logs with the use of OS library, so that the user does not have to do it manually

mirokovac commented on .gitignore:
```diff
@@ -1,2 +1,4 @@
 .idea/
```
@patrikprovaznik you need to manually remove files from project now that is pushed to github and add that to commit otherwise it will want to merge it with that... these .idea files can not be here.

mirokovac commented on task2/flask_app_config.ini:
```diff
@@ -0,0 +1,10 @@
+[flask]
+host = 127.0.0.1
+port = 5000
+secret_key = LsdGvcSYFDVP6b2hDhWJmA
```
@patrikprovaznik passwords in here needs to be only foo bar.. the key itself can t get to github cause anyone can misuse it. it s up to each person to create that password

mirokovac commented on task2/flask_app_config.ini:
```diff
@@ -0,0 +1,10 @@
+log_path_1 = /home/patrik/Documents/patprotasks/task2/logs/logger_1.log
+log_name = __name__
+log_level = logging.INFO
+app_debug = 1
```
@patrikprovaznik it is good practice to keep new line at the end of each file. (disclaimer - noone pays attention to such thing so it s not big mistake at all)

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,95 @@
+    if ssl:
+        return 'https'
+    else:
+        return 'http'
```
@patrikprovaznik you can do this in one line

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,95 @@
+        r = requests.get(url)
+        r.raise_for_status()
+    except Exception as e:
+        raise SystemExit(e)
```
@patrikprovaznik Log some error why the program stoped

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,95 @@
+        return logg_dict
+
+    def write_all_data(self) -> None:
+        output_folder = 'output_folder/'
```
@patrikprovaznik from arguments

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,95 @@
+        os.makedirs(output_folder, exist_ok=True)
+        with open(output_folder + 'output_data.json', 'w') as f:
+            f.write(self.interface_schema.dumps(self.parsed_data, many=True, ensure_ascii=False, indent=4))
+        with open(output_folder + 'log_data.json', 'w') as f:
```
@patrikprovaznik name of logger file from arguments or default that you ll have

mirokovac commented on task2/utils.py:
```diff
@@ -0,0 +1,23 @@
+                      'logging.WARNING': logging.WARNING,
+                      'logging.ERROR': logging.ERROR,
+                      }
+    set_log_level = log_level_info.get(log_level, logging.ERROR)
```
@patrikprovaznik you can do .get above when you are creating the dict.. no need to save the variable

mirokovac commented on task2/utils.py:
```diff
@@ -0,0 +1,23 @@
+    logger.setLevel(level=set_log_level)
+
+    formatter = logging.Formatter(fmt='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+    logs_folder = 'logs/'
```
@patrikprovaznik From arguments

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,95 @@
+
+class InterfaceManager:
+    def __init__(self, api_url, logger):
+        self.api_url = api_url
+        self.interface_schema = Interface.schema()
+        self.parsed_data = self._get_all_interfaces()
+        self.logger = logger
```
@patrikprovaznik make vars class protected with underscore

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,95 @@
+    def __init__(self, api_url, logger):
+        self.api_url = api_url
+        self.interface_schema = Interface.schema()
+        self.parsed_data = self._get_all_interfaces()
```
@patrikprovaznik please put this to separate method which makes it available to call it at any point.. I know I told you to have it in here only but at least app will make little more sense like that..

mirokovac reviewed (commented):
@patrikprovaznik make sure there is no .idea folder on review.. I think this is my biggest problem.. other things are just small adjustments

patrikprovaznik committed:
removed .idea folder

patrikprovaznik committed:
we did some final adjustments:
1. _get_all_interfaces is now being called from analyze and write_all_data methods and is no longer in __init__, so now we can call it at any point.
2. we got rid of function check_url and put it directly into _get_all_interfaces
3. creating output folder with files that belong there and log folder with name of logger can be now set by user from arguments
4. flask_app_config.ini is not in this task2 anymore, because we are not using it here and will be part of task3

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,88 @@
+        output_folder = self._args.out_fold
+        os.makedirs(output_folder, exist_ok=True)
+        with open(output_folder + self._args.out_data, 'w') as f:
+            f.write(self._interface_schema.dumps(parsed_data, many=True, ensure_ascii=False, indent=4))
```
@patrikprovaznik in files we usually don t indent to save on file size. especially since it is json which will probably be used only by some other app which will parse it anyway and there indentation does not matter

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,88 @@
+
+
+class InterfaceManager:
+    def __init__(self, api_url, logger, args):
```
@patrikprovaznik please use typehints here and everywhere where you have some parameters.. in some methods you are missing even typehint return type

mirokovac commented on task2/interface_manager.py:
```diff
@@ -0,0 +1,88 @@
+
+
+def main():
+    # Argument parser
+    args = arg_pars()
```
@patrikprovaznik sorry I didn t see this before.. but this is a bit tricky.. arguments should be in if __name__ == "__main__"
the reason for this is that you want to be able to use this function even outside this module. calling as a function from some other module.. this means that you will not have any arguments here but needs to be passed to this method.. 
please for passing arguments use *args or **kwargs.. (hint you need only one of them. to read more about this use https://www.geeksforgeeks.org/args-kwargs-python/), (hint2 be careful if arguments will exist - this needs to be checked)

mirokovac reviewed (changes requested):
@patrikprovaznik it starts to look very good. I have some issues I haven t seen before. the biggest is with arguments in main method. please fix. thank you

patrikprovaznik committed:
we did some final adjustments:
1. added typehints to all functions and methods where they were missing
2. removed indents from dumping json objects
3. we created main.py file, so now we can run interface_manager.py as module

mirokovac reviewed (changes requested):
@patrikprovaznik please apply CR as written in chat. Also I am missing requirements.txt file. when I tried to install this I was missing libraries..
the least you need there -- flask, marshmallow, dataclass_json and requests.

patrikprovaznik committed:
One final change has been made into code: now we can call main.py from another module without getting arguments from the command line, but now they are function parameters

mirokovac reviewed (approved):
@patrikprovaznik looks good
